src/main/java/org/broadinstitute/hellbender/utils/iterators/PushToPullIterator.java

```diff
@@ -0,0 +1,92 @@
+package org.broadinstitute.hellbender.utils.iterators;
+
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Iterator wrapper around our generic {@link PushPullTransformer)} interface. Wraps an iterator of input elements,
+ * and transforms the reads from that iterator using the provided transformer.
+ *
+ * Converts the push-style {@link PushPullTransformer)} interface to a pull model.
+ *
+ * @param <T> type of items to be iterated over
+ */
+public class PushToPullIterator<T> implements Iterator<T>, Iterable<T> {
+
+    private final Iterator<T> inputElements;
+    private final PushPullTransformer<T> transformer;
+    private Iterator<T> cachedElements = null;
+    private T nextElement = null;
+
+    /**
+     * @param inputElements wrapped iterator from which this iterator will pull elements
+     * @param transformer transformer through which the reads from the wrapped iterator will be fed
+     */
+    public PushToPullIterator(Iterator<T> inputElements, PushPullTransformer<T> transformer ) {
```

Comment:
The contract for these interfaces should probably state whether or not they're required to preserve input order. It seems like the `PushPullTransformer` is not required to, but the `PushToPullIterator` should (honor the order of the transformer) ?

Reply:
Good point. I have updated the javadoc to reflect this.